policy/src/main/java/org/apache/brooklyn/policy/location/CreateLocationPolicy.java

```diff
@@ -0,0 +1,288 @@
+
+import com.google.common.annotations.Beta;
+import com.google.common.base.Joiner;
+import com.google.common.base.Predicates;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.reflect.TypeToken;
+
+/**
+ * Policy that is attached to an entity to create a location configured
+ * using its sensor data. The created location is added to the catalog
+ * and will be removed when the entity is no longer running, based on
+ * the entity {@link Startable#SERVICE_UP} sensor status.
+ * <p>
+ * The policy definition includes configuration for the id, name and
+ * type of the location, as well as the {@link #LOCATION_CONFIG} map
+ * of config keys and the sensors used to define them.
+ * <p>
+ * The YAML below shows a policy that creates a {@code jclouds:aws-ec2}
+ * location with the region defined by a sensor on the entity it is
+ * attached to:
+ * <pre>{@code
+ * name: My App
+ * brooklyn.policies:
+ *   - type: org.apache.brooklyn.policy.location.CreateLocationPolicy
+ *     id: create-my-cloud-location
+ *     brooklyn.config:
+ *       location.catalogId: my-cloud
+ *       location.displayName: "My Cloud"
+ *       location.type: jclouds:aws-ec2
+ *       location.config:
+ *         region: $brooklyn:sensor("aws.region")
+ * }</pre>
+ *
+ */
+@Beta
+public class CreateLocationPolicy extends AbstractPolicy {
+
+    private static final Logger LOG = LoggerFactory.getLogger(CreateLocationPolicy.class);
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+        private String id;
+        private String name;
+        private AttributeSensor<Boolean> status;
+        private Map<String,AttributeSensor<?>> configuration;
+        private String catalogId;
+        private String displayName;
+        private String type;
+        private Set<String> tags;
+
+        public Builder id(String val) {
+            this.id = val; return this;
+        }
+        public Builder name(String val) {
+            this.name = val; return this;
+        }
+        public Builder status(AttributeSensor<Boolean> val) {
+            this.status = val; return this;
+        }
+        public Builder configuration(Map<String,AttributeSensor<?>> val) {
+            this.configuration = val; return this;
+        }
+        public Builder catalogId(String val) {
+            this.catalogId = val; return this;
+        }
+        public Builder displayName(String val) {
+            this.displayName = val; return this;
+        }
+        public Builder type(String val) {
+            this.type = val; return this;
+        }
+        public Builder tags(Set<String> val) {
+            this.tags = val; return this;
+        }
+        public CreateLocationPolicy build() {
+            return new CreateLocationPolicy(toFlags());
```

Comment:
Delete this `build` method - folk should always use the `PolicySpec`. I'd probably go further and delete the builder entirely, relying on using the standard `PolicySpec` as the way to build it.